관측지점정보

방재기상관측(AWS) 지점에 대한 지점번호, 관측 시작일과 종료일, 위·경도 등 이력 정보와 주변 환경 등에 대한 정보

In [1]:
import urllib.parse

base_url = "https://raw.githubusercontent.com/s0nghyunje0ng/2025-weather-bigdata-contest/safety/topic2_safety/data/external/"
encoded_filename = urllib.parse.quote("META_관측지점정보_20250610001229.csv")
url = base_url + encoded_filename

In [2]:
import pandas as pd

raw_meta = pd.read_csv(url, encoding="cp949")
raw_meta.head()

지점         시작일         종료일    지점명             지점주소          관리관서       위도  \
0  160  2020-10-23         NaN  부산(레)  부산광역시 서구 서대신동3가  부산지방기상청(159)  35.1188   
1  160  1993-03-04  2020-10-23  부산(레)  부산광역시 서구 서대신동3가  부산지방기상청(159)  35.1188   
2  296  2023-01-11  2025-04-25    북부산   부산광역시 강서구 대저1동  부산지방기상청(159)  35.2178   
3  296  2020-11-12  2023-01-11    북부산   부산광역시 강서구 대저1동  부산지방기상청(159)  35.2178   
4  904  2020-10-31         NaN     사상    부산광역시 사상구 삼락동  부산지방기상청(159)  35.1694   

         경도  노장해발고도(m)  기압계(관측장비지상높이(m))  기온계(관측장비지상높이(m))  풍속계(관측장비지상높이(m))  \
0  129.0000     512.00               NaN               NaN               NaN   
1  129.0000     518.46               NaN               NaN               NaN   
2  128.9602       3.00               NaN               NaN               NaN   
3  128.9602       3.00               NaN               NaN               NaN   
4  128.9751       0.73               NaN               NaN               NaN   

   강우계(관측장비지상높이(m))  
0               NaN  
1               NaN  
2               NaN  
3               NaN  
4               NaN

In [3]:
raw_meta.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65 entries, 0 to 64
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   지점                65 non-null     int64  
 1   시작일               65 non-null     object 
 2   종료일               53 non-null     object 
 3   지점명               65 non-null     object 
 4   지점주소              65 non-null     object 
 5   관리관서              65 non-null     object 
 6   위도                65 non-null     float64
 7   경도                65 non-null     float64
 8   노장해발고도(m)         65 non-null     float64
 9   기압계(관측장비지상높이(m))  0 non-null      float64
 10  기온계(관측장비지상높이(m))  0 non-null      float64
 11  풍속계(관측장비지상높이(m))  4 non-null      float64
 12  강우계(관측장비지상높이(m))  0 non-null      float64
dtypes: float64(7), int64(1), object(5)
memory usage: 6.7+ KB


2020년 5월 1일 ~ 2024년 10월 31일 사이에 관측이 이루어진 지점만 필터링

In [4]:
raw_meta["시작일"] = pd.to_datetime(raw_meta["시작일"])
raw_meta["종료일"] = pd.to_datetime(raw_meta["종료일"])

start_cutoff = pd.Timestamp("2020-05-01")
end_cutoff = pd.Timestamp("2024-10-31")

active_meta = raw_meta[(raw_meta["시작일"] <= end_cutoff) & ((raw_meta["종료일"].isna()) | (raw_meta["종료일"] >= start_cutoff))]
active_meta

지점        시작일        종료일    지점명               지점주소          관리관서  \
0   160 2020-10-23        NaT  부산(레)    부산광역시 서구 서대신동3가  부산지방기상청(159)   
1   160 1993-03-04 2020-10-23  부산(레)    부산광역시 서구 서대신동3가  부산지방기상청(159)   
2   296 2023-01-11 2025-04-25    북부산     부산광역시 강서구 대저1동  부산지방기상청(159)   
3   296 2020-11-12 2023-01-11    북부산     부산광역시 강서구 대저1동  부산지방기상청(159)   
4   904 2020-10-31        NaT     사상      부산광역시 사상구 삼락동  부산지방기상청(159)   
5   904 2019-12-10 2020-10-31     사상      부산광역시 사상구 감전동  부산지방기상청(159)   
10  910 2024-10-15        NaT     영도      부산광역시 영도구 동삼동  부산지방기상청(159)   
11  910 2020-11-01 2024-10-15     영도      부산광역시 영도구 동삼동  부산지방기상청(159)   
12  910 2020-10-23 2020-11-01     영도      부산광역시 영도구 동삼동  부산지방기상청(159)   
13  910 2010-07-23 2020-10-23     영도      부산광역시 영도구 동삼동  부산지방기상청(159)   
15  921 2024-10-13        NaT    가덕도      부산광역시 강서구 대항동  부산지방기상청(159)   
16  921 2020-10-23 2024-10-13    가덕도      부산광역시 강서구 대항동  부산지방기상청(159)   
17  921 2011-08-03 2020-10-23    가덕도      부산광역시 강서구 대항동  부산지방기상청(159)   
19  923 2021-07-23        NaT     기장  부산광역시 기장군 장안읍 길천리  부산지방기상청(159)   
20  923 2020-10-23 2021-07-23     기장  부산광역시 기장군 일광면 이천리  부산지방기상청(159)   
21  923 2011-09-22 2020-10-23     기장  부산광역시 기장군 일광면 이천리  부산지방기상청(159)   
24  937 2023-11-01 2024-11-30    해운대      부산광역시 해운대구 우동  부산지방기상청(159)   
25  937 2020-10-23 2023-11-01    해운대      부산광역시 해운대구 우동  부산지방기상청(159)   
26  937 2010-06-29 2020-10-23    해운대      부산광역시 해운대구 우동  부산지방기상청(159)   
29  938 2020-10-15        NaT    부산진     부산광역시 부산진구 개금동  부산지방기상청(159)   
30  938 2012-12-07 2020-10-15    부산진     부산광역시 부산진구 개금동  부산지방기상청(159)   
35  939 2020-10-14        NaT    금정구      부산광역시 금정구 두구동  부산지방기상청(159)   
36  939 2020-06-12 2020-10-14    금정구      부산광역시 금정구 두구동  부산지방기상청(159)   
37  939 2014-04-03 2020-06-12    금정구      부산광역시 금정구 두구동  부산지방기상청(159)   
41  940 2022-10-14        NaT     동래      부산광역시 동래구 복천동  부산지방기상청(159)   
42  940 2014-11-14 2022-10-14     동래      부산광역시 동래구 복천동  부산지방기상청(159)   
47  941 2020-10-23        NaT     북구       부산광역시 북구 덕천동  부산지방기상청(159)   
48  941 2015-10-01 2020-10-23     북구       부산광역시 북구 덕천동  부산지방기상청(159)   
51  942 2021-07-22        NaT   부산남구       부산광역시 남구 감만동  부산지방기상청(159)   
52  942 2020-10-23 2021-07-22   부산남구       부산광역시 남구 감만동  부산지방기상청(159)   
53  942 2019-04-27 2020-10-23   부산남구       부산광역시 남구 감만동  부산지방기상청(159)   
56  950 2023-11-01        NaT     사하      부산광역시 사하구 신평동  부산지방기상청(159)   
57  950 2011-12-08 2023-11-01     사하      부산광역시 사하구 신평동  부산지방기상청(159)   
62  969 2017-12-06 2020-07-27     북항       부산광역시 남구 용호동  부산지방기상청(159)   

         위도        경도  노장해발고도(m)  기압계(관측장비지상높이(m))  기온계(관측장비지상높이(m))  \
0   35.1188  129.0000     512.00               NaN               NaN   
1   35.1188  129.0000     518.46               NaN               NaN   
2   35.2178  128.9602       3.00               NaN               NaN   
3   35.2178  128.9602       3.00               NaN               NaN   
4   35.1694  128.9751       0.73               NaN               NaN   
5   35.1370  128.9675       4.28               NaN               NaN   
10  35.0661  129.0742     138.96               NaN               NaN   
11  35.0661  129.0742     138.96               NaN               NaN   
12  35.0661  129.0742     138.96               NaN               NaN   
13  35.0661  129.0742     137.94               NaN               NaN   
15  34.9931  128.8314      75.63               NaN               NaN   
16  34.9931  128.8314      75.63               NaN               NaN   
17  34.9931  128.8314      73.26               NaN               NaN   
19  35.2751  129.2511      65.99               NaN               NaN   
20  35.2751  129.2511      65.99               NaN               NaN   
21  35.2750  129.2511      65.00               NaN               NaN   
24  35.1761  129.1624      63.72               NaN               NaN   
25  35.1761  129.1624      63.72               NaN               NaN   
26  35.1761  129.1624      62.98               NaN               NaN   
29  35.15

In [5]:
# 지점별로 위도와 경도에 대한 평균과 표준편차 계산
coord_stats = active_meta.groupby("지점")[["위도", "경도"]].agg(["mean", "std"]).reset_index()
coord_stats

지점         위도                    경도          
              mean       std        mean       std
0   160  35.118800  0.000000  129.000000  0.000000
1   296  35.217800  0.000000  128.960200  0.000000
2   904  35.153200  0.022910  128.971300  0.005374
3   910  35.066100  0.000000  129.074200  0.000000
4   921  34.993100  0.000000  128.831400  0.000000
5   923  35.275067  0.000058  129.251100  0.000000
6   937  35.176100  0.000000  129.162400  0.000000
7   938  35.158900  0.000000  129.019250  0.000071
8   939  35.293200  0.000000  129.103500  0.000000
9   940  35.209100  0.000000  129.090100  0.000000
10  941  35.213000  0.000000  129.002600  0.000000
11  942  35.118600  0.000000  129.088367  0.000058
12  950  35.090000  0.000000  128.983900  0.000000
13  969  35.091000       NaN  129.127000       NaN

관측소 위도/경도 표준편차가 거의 없음 → 같은 자리에서 시작/종료를 반복했을 것으로 추정

In [6]:
# 지점별 위도/경도 평균 테이블
coord_mean = active_meta.groupby("지점")[["위도", "경도"]].mean().reset_index()

# 지점, 지점명, 지점주소 정보 중복 제거 후 하나로 정리
station_info = active_meta.drop_duplicates("지점")[["지점", "지점명", "지점주소"]]

# 병합하여 위도/경도 평균 추가
stations_avg_coord = pd.merge(station_info, coord_mean, on="지점", how="left")
stations_avg_coord

지점    지점명               지점주소         위도          경도
0   160  부산(레)    부산광역시 서구 서대신동3가  35.118800  129.000000
1   296    북부산     부산광역시 강서구 대저1동  35.217800  128.960200
2   904     사상      부산광역시 사상구 삼락동  35.153200  128.971300
3   910     영도      부산광역시 영도구 동삼동  35.066100  129.074200
4   921    가덕도      부산광역시 강서구 대항동  34.993100  128.831400
5   923     기장  부산광역시 기장군 장안읍 길천리  35.275067  129.251100
6   937    해운대      부산광역시 해운대구 우동  35.176100  129.162400
7   938    부산진     부산광역시 부산진구 개금동  35.158900  129.019250
8   939    금정구      부산광역시 금정구 두구동  35.293200  129.103500
9   940     동래      부산광역시 동래구 복천동  35.209100  129.090100
10  941     북구       부산광역시 북구 덕천동  35.213000  129.002600
11  942   부산남구       부산광역시 남구 감만동  35.118600  129.088367
12  950     사하      부산광역시 사하구 신평동  35.090000  128.983900
13  969     북항       부산광역시 남구 용호동  35.091000  129.127000

In [7]:
stations_avg_coord["지점"].unique()

array([160, 296, 904, 910, 921, 923, 937, 938, 939, 940, 941, 942, 950,
       969])

우리가 가진 `stn`과 겹치지 않는 지점 3개 발견 ❗ `160`, `296`, `969`

In [8]:
active_meta[active_meta["지점"].isin([160, 296, 969])]

지점        시작일        종료일    지점명             지점주소          관리관서       위도  \
0   160 2020-10-23        NaT  부산(레)  부산광역시 서구 서대신동3가  부산지방기상청(159)  35.1188   
1   160 1993-03-04 2020-10-23  부산(레)  부산광역시 서구 서대신동3가  부산지방기상청(159)  35.1188   
2   296 2023-01-11 2025-04-25    북부산   부산광역시 강서구 대저1동  부산지방기상청(159)  35.2178   
3   296 2020-11-12 2023-01-11    북부산   부산광역시 강서구 대저1동  부산지방기상청(159)  35.2178   
62  969 2017-12-06 2020-07-27     북항     부산광역시 남구 용호동  부산지방기상청(159)  35.0910   

          경도  노장해발고도(m)  기압계(관측장비지상높이(m))  기온계(관측장비지상높이(m))  풍속계(관측장비지상높이(m))  \
0   129.0000     512.00               NaN               NaN               NaN   
1   129.0000     518.46               NaN               NaN               NaN   
2   128.9602       3.00               NaN               NaN               NaN   
3   128.9602       3.00               NaN               NaN               NaN   
62  129.1270      37.72               NaN               NaN               NaN   

    강우계(관측장비지상높이(m))  
0                NaN  
1                NaN  
2                NaN  
3                NaN  
62               NaN

- `160`: `159`와 겹침 → 대체

- `296`: `904`와 겹침 → 제거

- `969`: 더이상 사용하지 않는 지점 → 제거

In [9]:
stations_avg_coord["지점"] = stations_avg_coord["지점"].replace(160, 159)
stations_final = stations_avg_coord[~stations_avg_coord["지점"].isin([296, 969])].reset_index(drop=True)
stations_final

지점    지점명               지점주소         위도          경도
0   159  부산(레)    부산광역시 서구 서대신동3가  35.118800  129.000000
1   904     사상      부산광역시 사상구 삼락동  35.153200  128.971300
2   910     영도      부산광역시 영도구 동삼동  35.066100  129.074200
3   921    가덕도      부산광역시 강서구 대항동  34.993100  128.831400
4   923     기장  부산광역시 기장군 장안읍 길천리  35.275067  129.251100
5   937    해운대      부산광역시 해운대구 우동  35.176100  129.162400
6   938    부산진     부산광역시 부산진구 개금동  35.158900  129.019250
7   939    금정구      부산광역시 금정구 두구동  35.293200  129.103500
8   940     동래      부산광역시 동래구 복천동  35.209100  129.090100
9   941     북구       부산광역시 북구 덕천동  35.213000  129.002600
10  942   부산남구       부산광역시 남구 감만동  35.118600  129.088367
11  950     사하      부산광역시 사하구 신평동  35.090000  128.983900

인접 관측소 매핑

In [10]:
from geopy.distance import geodesic

# 인접 지점 매핑 (각 지점별 가장 가까운 3개 지점 찾기)
neighbor_rows = []

for idx, row in stations_final.iterrows():
    origin_id = row["지점"]
    origin_name = row["지점명"]
    origin_coord = (row["위도"], row["경도"])

    distances = []

    for idx2, row2 in stations_final.iterrows():
        if row["지점"] == row2["지점"]:
            continue
        target_id = row2["지점"]
        target_name = row2["지점명"]
        target_coord = (row2["위도"], row2["경도"])
        distance_km = geodesic(origin_coord, target_coord).km
        distances.append((target_id, target_name, distance_km))

    # 거리순 정렬 후 상위 3개만 저장
    nearest = sorted(distances, key=lambda x: x[2])[:3]
    for neighbor_id, neighbor_name, dist in nearest:
        neighbor_rows.append({
            "지점": origin_id,
            "지점명": origin_name,
            "인접지점": neighbor_id,
            "인접지점명": neighbor_name,
            "거리(km)": round(dist, 4)
        })

neighbors_df = pd.DataFrame(neighbor_rows)
neighbors_df

지점    지점명  인접지점  인접지점명   거리(km)
0   159  부산(레)   950     사하   3.5162
1   159  부산(레)   904     사상   4.6267
2   159  부산(레)   938    부산진   4.7822
3   904     사상   938    부산진   4.4145
4   904     사상   159  부산(레)   4.6267
5   904     사상   950     사하   7.1050
6   910     영도   942   부산남구   5.9660
7   910     영도   950     사하   8.6518
8   910     영도   159  부산(레)   8.9421
9   921    가덕도   950     사하  17.5834
10  921    가덕도   159  부산(레)  20.7615
11  921    가덕도   904     사상  21.8700
12  923     기장   939    금정구  13.5773
13  923     기장   937    해운대  13.6294
14  923     기장   940     동래  16.3799
15  937    해운대   940     동래   7.5340
16  937    해운대   942   부산남구   9.2847
17  937    해운대   938    부산진  13.1801
18  938    부산진   904     사상   4.4145
19  938    부산진   159  부산(레)   4.7822
20  938    부산진   941     북구   6.1907
21  939    금정구   940     동래   9.4098
22  939    금정구   941     북구  12.7863
23  939    금정구   923     기장  13.5773
24  940     동래   937    해운대   7.5340
25  940     동래   941     북구   7.9789
26  940     동래   938    부산진   8.5242
27  941     북구   938    부산진   6.1907
28  941     북구   904     사상   7.2211
29  941     북구   940     동래   7.9789
30  942   부산남구   910     영도   5.9660
31  942   부산남구   938    부산진   7.7244
32  942   부산남구   159  부산(레)   8.0552
33  950     사하   159  부산(레)   3.5162
34  950     사하   904     사상   7.1050
35  950     사하   938    부산진   8.2953

In [11]:
neighbors_dict = neighbors_df.groupby("지점")["인접지점"].apply(list).to_dict()
neighbors_dict

{159: [950, 904, 938],
 904: [938, 159, 950],
 910: [942, 950, 159],
 921: [950, 159, 904],
 923: [939, 937, 940],
 937: [940, 942, 938],
 938: [904, 159, 941],
 939: [940, 941, 923],
 940: [937, 941, 938],
 941: [938, 904, 940],
 942: [910, 938, 159],
 950: [159, 904, 938]}

In [12]:
import json

# 저장
with open("busan_stn_neighbors.json", "w", encoding="utf-8") as f:
    json.dump(neighbors_dict, f, ensure_ascii=False)